paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/PaparazziPlugin.kt

```diff
@@ -166,62 +166,67 @@ public class PaparazziPlugin @Inject constructor(
-          test.systemProperties["paparazzi.layoutlib.runtime.root"] =
-            layoutlibNativeRuntimeFileCollection.singleFile.absolutePath
-          test.systemProperties["paparazzi.layoutlib.resources.root"] =
-            layoutlibResourcesFileCollection.singleFile.absolutePath
-          test.systemProperties["paparazzi.test.record"] = isRecordRun.get()
-          test.systemProperties["paparazzi.test.verify"] = isVerifyRun.get()
-        }
-
-        test.doLast {
-          val uri = reportOutputDir.get().asFile.toPath().resolve("index.html").toUri()
-          test.logger.log(LIFECYCLE, "See the Paparazzi report at: $uri")
+        val isRecordRun = recordTaskProvider.map { graph.hasTask(it) }.get()
+        val isVerifyRun = verifyTaskProvider.map { graph.hasTask(it) }.get()
+
+        testTaskProvider.configureEach { test ->
+          test.systemProperties["paparazzi.test.resources"] =
+            writeResourcesTask.flatMap { it.paparazziResources.asFile }.get().path
+          test.systemProperties["paparazzi.project.dir"] = projectDirectory.toString()
+          test.systemProperties["paparazzi.build.dir"] = buildDirectory.get().toString()
+          test.systemProperties["paparazzi.report.dir"] = reportOutputDir.get().toString()
+          test.systemProperties["paparazzi.snapshot.dir"] = snapshotOutputDir.toString()
+          test.systemProperties["paparazzi.artifacts.cache.dir"] = gradleUserHomeDir.path
+          test.systemProperties.putAll(project.properties.filterKeys { it.startsWith("app.cash.paparazzi") })
+
+          test.inputs.property("paparazzi.test.record", isRecordRun)
+          test.inputs.property("paparazzi.test.verify", isVerifyRun)
+
+          test.inputs.files(sources.localResourceDirs)
+            .withPropertyName("paparazzi.localResourceDirs")
+            .withPathSensitivity(PathSensitivity.RELATIVE)
+          test.inputs.files(sources.moduleResourceDirs)
+            .withPropertyName("paparazzi.moduleResourceDirs")
+            .withPathSensitivity(PathSensitivity.RELATIVE)
+          test.inputs.files(sources.localAssetDirs)
+            .withPropertyName("paparazzi.localAssetDirs")
+            .withPathSensitivity(PathSensitivity.RELATIVE)
+          test.inputs.files(sources.moduleAssetDirs)
+            .withPropertyName("paparazzi.moduleAssetDirs")
+            .withPathSensitivity(PathSensitivity.RELATIVE)
+          test.inputs.files(layoutlibNativeRuntimeFileCollection)
+            .withPropertyName("paparazzi.nativeRuntime")
+            .withPathSensitivity(PathSensitivity.NONE)
+
+          test.outputs.dir(reportOutputDir)
+
+          if (isVerifyRun) {
+            test.inputs.dir(snapshotOutputDir)
+              .withPropertyName("paparazzi.snapshot.output.dir")
+              .withPathSensitivity(PathSensitivity.RELATIVE)
+          }
+          if (isRecordRun) {
+            test.outputs.dir(snapshotOutputDir)
+          }
```

Comment:
discussed offline on Slack, but sharing here as well:

Given that the Gradle build tool team raised some potential concerns about configuring the entire Test task in the  `TaskExecutionGraph.whenReady` block (i.e., missing any inputs that might affect the graph, like task dependencies), maybe a better alternative would be deferring the resolution of `isVerifyRun`/`isRecordRun` as suggested here: https://github.com/cashapp/paparazzi/pull/852#discussion_r1194683002, i.e., 

```
        val dummyUnusedDir = project.layout.projectDirectory

        test.inputs.dir(isVerifyRun.map { if (it) snapshotOutputDir else dummyUnusedDir })
          .withPropertyName("paparazzi.snapshot.input.dir")
          .withPathSensitivity(PathSensitivity.RELATIVE)

        test.outputs.dir(isRecordRun.map { if (it) snapshotOutputDir else dummyUnusedDir })
          .withPropertyName("paparazzi.snapshot.output.dir")
```

(my choice of `dummyUnusedDir` might be wrong, feel free to suggest an alternative, but you get the gist...)

Reply:
another suggestion would be to take what you provided @~jrod and make these input/output properties optional since the else case uses a dummyDir.

Reply:
@jrodbx I'm also not really a fan of a dummy unused dir, as those inputs/outputs are still being tracked and I believe it only adds to the confusion, I believe not defining them when not needed is a cleaner solution. We could also do task configuration in two steps: the "static" one, where most of task is configured, except for stuff that relies on taskGraph, and the "dynamic" part, where the mode is decided and all the configuration tied to it is applied.

Reply:
> We could also do task configuration in two steps: the "static" one, where most of task is configured, except for stuff that relies on taskGraph, and the "dynamic" part, where the mode is decided and all the configuration tied to it is applied.

I find dummyUnusedDir to be more clear than dividing into "static" and "dynamic" configuration, which imo is harder for maintainers to contribute/grok the intent of the configuration.

>  I believe not defining them when not needed is a cleaner solution. 

If there is a way to take @geoff-powell's proposal and make those properties optional, then i'm all for that! However, I struggle to see how this can be accomplished conditionally, while still deferring the isVerifyRun/isRecordRun resolution...in other words,

```
     if (isVerifyRun.get()) { 
        test.inputs.dir(snapshotOutputDir)
          .optional()
          .withPropertyName("paparazzi.snapshot.input.dir")
          .withPathSensitivity(PathSensitivity.RELATIVE)
     }
     
     ...
```

still won't work, because of the race condition between whenReady and test task configuration!

This is likely why `@TWiStErRob` suggested (and I agree with) the dummy unused dir approach.  The point of the exercise here to defer resolution of isVerifyRun/isRecordRun.  Non-ideal?  Perhaps, but let's fix the current issue and brainstorm a better design later.  Based on our previous discussions, a more ideal solution involves a huge redesign/refactoring in order to move away from whenReady, but we really just want to solve the caching issue now for our consumers.

Reply:
Applied the fix suggested, this works much better than dummy dir 👍